### PR TITLE
Implement efficient cluster state snapshots in ECSStateClient

### PR DESCRIPTION
**Pull request by wjbuys**
This uses an `ECSAsyncClient` to implement the `ECSState` interface, and updates the `SingleTaskScheduler` to actually only run a single task.

This also changes the types in the snapshot to be lists of instances/tasks, since the maps weren't very useful in practice.

The integration test is very basic, but we may be able to build on it for further integration tests that require a test cluster.

Tracking issue: #261

#### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

### CODE REVIEW

**Comment by wjbuys** [comment]
Looks like there's another ECS client getting started in the Spring tests, modified the build logic to show more failure details in Travis to see what's going on.

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/state/FakeECSStateClient.java
Comment (wjbuys) [inline scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/state/FakeECSStateClient.java]: That's weird, I thought I deleted this file. I'll remove it.

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java
Comment (narehayrapetyan) [inline scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java:34]: would it be possible to put all the configuration in one package so it's easy to find?

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java
Comment (narehayrapetyan) [inline scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java]: you'll have to create a new internal object for environmentversion or something similar because the returned type is now describeEnvironment which contains separate fields. can be done in the refactoring cr

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java
Comment (narehayrapetyan) [inline scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java:40]: do we need to sort? I suppose this is gonna be removed anyway

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/StartTask.java
Comment (narehayrapetyan) [inline scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/StartTask.java:43]: constant

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/state/PaginatedResponseIterator.java
Comment (narehayrapetyan) [inline scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/state/PaginatedResponseIterator.java:32]: why are these classes not public?

File: scheduling-manager/src/test/java/com/amazonaws/blox/scheduling/state/ContainerInstanceListerTest.java
Comment (narehayrapetyan) [inline scheduling-manager/src/test/java/com/amazonaws/blox/scheduling/state/ContainerInstanceListerTest.java:59]: is this one of the temporary unit/integ mix tests? can we move them to a separate package so they're easy to refactor into integ tests later?

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java
Comment (wjbuys) [reply scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java:34]: In this case, it's not the configuration that's in a different package, it's the `@Component` annotation on the state stuff. I'm open to any suggestions on how to deal with that.

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java
Comment (wjbuys) [reply scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java]: That's what I did for #278

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java
Comment (wjbuys) [reply scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/SingleTaskScheduler.java:40]: yes. Just so we don't start the task multiple times on different instances just because the ordering is different. (I had that before, and ended up with a cluster full of random tasks)

File: scheduling-manager/src/test/java/com/amazonaws/blox/scheduling/state/ContainerInstanceListerTest.java
Comment (wjbuys) [reply scheduling-manager/src/test/java/com/amazonaws/blox/scheduling/state/ContainerInstanceListerTest.java:59]: No, this unit test has value, and will need to be expanded to cover failure cases that are hard/impossible to reproduce against a real ECS client (e.g. partial asynchronous failure)

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/state/PaginatedResponseIterator.java
Comment (wjbuys) [reply scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/state/PaginatedResponseIterator.java:32]: Think it was left over from extracting this as a private inner class on one of the listers. Should it be public? I do not have strong feelings about java class visibility beyond what's needed to make the compiler happy, so I'll defer to whatever your guideline is here.

**Comment by wjbuys** [comment]
Just to follow up, I extracted some of this PR into #278 to fix the build, so I'll have to rebase this PR after that one is merged to `dev`.

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/StartTask.java
Comment (wjbuys) [reply scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/engine/StartTask.java:43]: It's a single usage of an unimportant string in a temporary class. I'll extract it, but I'm not convinced it's helpful.

**Comment by wjbuys** [comment]
I've rebased this PR onto dev, should be ready to review/ship now.

File: scheduling-manager/src/test/java/com/amazonaws/blox/scheduling/state/ContainerInstanceListerTest.java
Comment (wjbuys) [reply scheduling-manager/src/test/java/com/amazonaws/blox/scheduling/state/ContainerInstanceListerTest.java:59]: Oh, and to answer the second part of your question: They are pure unit tests, since they mock out all dependencies, do I don't think they belong in a separate package.

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java
Comment (narehayrapetyan) [reply scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java:34]: What do you mean? I think the schedulingapplication configuration is not in com...manager

**Review by narehayrapetyan — APPROVED** [review]

File: scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java
Comment (wjbuys) [reply scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/manager/ManagerApplication.java:34]: Oh, wait, dammit github highlighting. Do you mean "why is SchedulingApplication in a different package"? If so, the reason is that it's the shared config between all components of the scheduling-manager package. If I were to put it in any of the subpackages, then it would no longer be clear that it's common to all of them.

An alternative could be to add a separate top-level `com.amazonaws.blox.scheduling.config` package, and add all the config in there. Is that what you were alluding to?

**Pull request merged**
